### ISSUE TITLE
TF 2.3 training slowed down by 15% compared to 2.2

### ISSUE THREAD

**Issue by lgeiger** [tensorflow/tensorflow#41827]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.3.0, 2.4.0.dev20200728
- Python version: 3.7.8
- CUDA/cuDNN version: 10.1 / 7.6.5.32
- GPU model and memory: NVIDIA V100 on 12 vCPUs, 40 GB memory GCP node

**Describe the current behavior**

When upgrading from TensorFlow 2.2.0 to 2.3.0 we observed a 15 - 18% slow down in training speed for our workloads. Unfortunately I wasn't able to find an easy to reproduce example before the stable release was cut, but below is a code example that illustrates the performance degradation.

When running the training script on a single NVIDIA V100 a 15% performance loss compared to 2.2 can be observed which still is noticable in the latest nightly:

version | epoch time | step time | GPU idle time
--- | ---  | --- | ---
2.2.0 | 34 s | 124.3 ms | 19.7 ms (15.6 %)
2.3.0 | 39 s | 141.9 ms | 37.2 ms (26.1 %)
2.4.0.dev20200728 | 38s | 136.2 ms | 31.6 ms (23.2 %)

**On Device: total self-time (grouped by type)**
2.2.0 | 2.3.0 | 2.4.0.dev20200728
--- | --- | ---
<img width="395" alt="Screenshot 2020-07-28 at 17 45 02" src="https://user-images.githubusercontent.com/13285808/88689213-80d3ff80-d0fa-11ea-88ec-3feda0cf9c24.png"> | <img width="375" alt="Screenshot 2020-07-28 at 17 45 15" src="https://user-images.githubusercontent.com/13285808/88689219-829dc300-d0fa-11ea-99bb-5d406805a6e8.png"> | <img width="359" alt="Screenshot 2020-07-28 at 17 46 08" src="https://user-images.githubusercontent.com/13285808/88689223-83365980-d0fa-11ea-91aa-46ec7659fe64.png">

The example uses auto mixed precision, but the slowdown can also be observed when running in `float32` or in multi-GPU training. When looking at the [generated execution profile](https://github.com/tensorflow/tensorflow/files/4989125/tb-profile.zip) the slowdown can be explained by an increased idle time of the GPU. Since the training data is cached in memory there should be no IO bottleneck so I am not sure if this performance regression is caused by `tf.data` or by the runtime itself.

**Describe the expected behavior**

TensorFlow 2.3 should show equally fast training performance compared to 2.2.

**Standalone code to reproduce the issue**
```python
import tensorflow as tf
import tensorflow_datasets as tfds

batch_size = 64


def _decode_and_center_crop(image_bytes):
    """Crops to center of image with padding then scales image_size."""
    shape = tf.image.extract_jpeg_shape(image_bytes)
    image_height, image_width, image_size = shape[0], shape[1], 224

    padded_center_crop_size = tf.cast(
        (
            (image_size / (image_size + 32))
            * tf.cast(tf.minimum(image_height, image_width), tf.float32)
        ),
        tf.int32,
    )

    offset_height = ((image_height - padded_center_crop_size) + 1) // 2
    offset_width = ((image_width - padded_center_crop_size) + 1) // 2
    crop_window = tf.stack(
        [offset_height, offset_width, padded_center_crop_size, padded_center_crop_size]
    )
    image = tf.image.decode_and_crop_jpeg(image_bytes, crop_window, channels=3)
    return tf.image.resize(image, [image_size, image_size], method="bicubic")


def preprocessing(data):
    return (
        tf.cast(_decode_and_center_crop(data["image"]), tf.float32),
        data["label"],
    )


dataset = tfds.load(
    "imagenette", decoders={"image": tfds.decode.SkipDecoding()}, split="train",
)

dataset = (
    dataset.cache()
    .repeat(2)  # Artificially increase time per epoch to make it easier to measure
    .map(preprocessing, num_parallel_calls=tf.data.experimental.AUTOTUNE)
    .batch(batch_size)
    .prefetch(1)
)

with tf.distribute.MirroredStrategy().scope():
    model = tf.keras.applications.ResNet50(weights=None)

    model.compile(
        optimizer=tf.train.experimental.enable_mixed_precision_graph_rewrite(
            tf.keras.optimizers.Adam(), loss_scale="dynamic"
        ),
        loss="sparse_categorical_crossentropy",
    )

tb_cbk = tf.keras.callbacks.TensorBoard(f"logs/{tf.__version__}", profile_batch=300)
model.fit(dataset, verbose=2, epochs=3, callbacks=[tb_cbk])
```

**Other info / logs**

TensorBoard profiles for the runs mentioned above are available at [tb-profile.zip](https://github.com/tensorflow/tensorflow/files/4989125/tb-profile.zip)

@mihaimaruseac @jsimsa @guptapriya do you mind taking a look at this?

**Comment by geetachavan1** [comment]
I was able to reproduce the issue. Here is the [gist](https://colab.research.google.com/gist/geetachavan1/6208ca2c63fe8d3f58d0928b23585d0b/untitled.ipynb#scrollTo=8KXFXJIsBjS-)..

**Comment by guptapriya** [comment]
Thanks for the report @lgeiger . We will look more into this. The code sample uses MirroredStrategy - so I wanted to clarify when you said this regression is noticed on a single GPU as well - is that with or without MirroredStrategy? If latter, we would start investigating that case first (i.e. 1 GPU, no distribution, no mixed precision).

**Comment by lgeiger** [comment]
@guptapriya Thanks for looking into this.

> The code sample uses MirroredStrategy - so I wanted to clarify when you said this regression is noticed on a single GPU as well - is that with or without MirroredStrategy?

Sorry about that, I missed that since I copied the example from a past issue I had with multi-GPU. I ran a few more benchmarks with the above example on a single GPU machine:

version | FP32 | FP32 Mirrored | FP16 | FP16 Mirrored
--- | --- | --- | --- | ---
2.2.0 | 55 s | 55 s | 36s | 34 s
2.3.0 | 54 s | **58 s** | 37 s | **39 s**

Indeed, it looks like whether mirrored strategy is used or not has a large influence. I am not sure why there is a difference in execution speed for mixed precision with and without a strategy, altough I think that might be a seperate issue.

One thing to note is that 2.3 logs the following deprecation warning when used with mirrored strategy which wasn't present before, but that might be unrelated as well:
```
WARNING:tensorflow:From ~/.local/lib/python3.7/site-packages/tensorflow/python/data/ops/multi_device_iterator_ops.py:601: get_next_as_optional (from tensorflow.python.data.ops.iterator_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.data.Iterator.get_next_as_optional()` instead.
```

Note that this slowdown is more noticable with mixed precision as the kernel execution time is smaller so the increased idle time is easier to spot.

**Comment by guptapriya** [comment]
Thanks Lukas for the update, that helps a lot. I can confirm that we have verified the regression (looks like it happened sometime back in April. 
We will update here when we have info into the root cause and fixes.

**Comment by lgeiger** [comment]
@guptapriya Thank you very much the fast response, I am glad that you were able to verify the regression.

> I can confirm that we have verified the regression (looks like it happened sometime back in April)

April sounds like a long time for the regression to stay unnoticed. Am I doing something abnormal in the example or am I missing a best practice here?

**Comment by guptapriya** [comment]
Actually now we are no longer sure about the timing because the benchmark I used to get the timing was actually changed in April. So we are now looking into the timing of regression again. It may have come up later. 
Once we determine the root cause, we should be able to tell if there is anything in your code that makes this show up or simply on TF side that somehow went unnoticed.
In your experience, when did you start noticing the regression?

**Comment by lgeiger** [comment]
Makes sense, thanks for the help.

>  In your experience, when did you start noticing the regression?

I usually stay on the stable versions of TensorFlow for production workloads so I first noticed it around 2.3 RC 0 or 1 when running our internal ImageNet sanity check before upgrading TF version. But it took some time to find an reliably reproducible example that I could use to open the issue.

**Comment by zongweiz** [comment]
@lgeiger thanks for providing the tensorboard profile. We have some observations regarding the profile which use fp16. You mentioned that the regression also happens on fp32 (and multi-gpu). Do you happen to have the fp32 profiles of TF 2.2 and 2.3? Thanks.

**Comment by lgeiger** [comment]
@zongweiz Sure, I reran the above code example (float32 with mirrored strategy) to generate these profiles: 
[tb-profile-fp32.zip](https://github.com/tensorflow/tensorflow/files/4997769/tb-profile-fp32.zip)

**Comment by lgeiger** [comment]
I hope the profiles are helpful for debugging, do you have any news on the resolution of this issue?

**Comment by guptapriya** [comment]
@jaingaurav is looking into a potential fix I believe.

**Comment by jaingaurav** [comment]
Yes, we found some unintended host to device copies caused by a previous change, that I am trying to eliminate.

**Comment by dathudeptrai** [comment]
i also found that tensorflow 2.3 mixed precision didn't speed-up training as tensorflow 2.2.0 does :)).

**Comment by lgeiger** [comment]
> Yes, we found some unintended host to device copies caused by a previous change, that I am trying to eliminate.

@jaingaurav Thanks for looking into it. Do you know if this fix will make it into the 2.4 release?

**Comment by jaingaurav** [comment]
@lgeiger: Yes this is currently planned for the 2.4 release. However the fix is still being worked on.

**Comment by lgeiger** [comment]
@jaingaurav I just checked with the new release candidate and CUDA 11, but unfortunately this issue still exists and the increased idle time is clearly visible in the profiles.

Below are the numbers for `2.4.0rc0` including the [profiles for the runs](https://github.com/tensorflow/tensorflow/files/5478702/logs.zip):

version | FP32 | FP32 Mirrored | FP16 | FP16 Mirrored
--- | --- | --- | --- | ---
2.4.0rc0 | 54s | **60s** | 40s | **41s**

**Comment by jaingaurav** [comment]
cc: @rohan100jain, @zongweiz looks like the fix in 673b993 didn't quite work as well as we'd hoped.

**Comment by lgeiger** [comment]
Thanks for looking into it. I also updated the table with measurements using the new [Keras mixed precision API](https://github.com/tensorflow/community/pull/293) which makes the slowdown easier to spot.

**Comment by lgeiger** [comment]
@rohan100jain Has there been any progress on this? From briefly skimming the changes in RC1, it looks like it doesn't include a fix for this.

This bug has been blocking me from upgrading to 2.3 so it would be great to get this resolved in 2.4. Please let me know if there anything that I could do from my side to help you debug this further?

**Comment by lgeiger** [comment]
I double checked and the issue still exists in tf-nightly 2.5.0-dev20201110.

**Comment by lgeiger** [comment]
@rohan100jain @zongweiz @jaingaurav I did a bit more testing with `2.4.0rc1` and it looks like when using `.batch(batch_size, drop_remainder=True)` instead of `.batch(batch_size)`, both mirrored strategy and normal training behave the same in float32 and the slowdown is significantly reduced for mixed precision training.

I am a bit confused why keeping the remainder results in such a significant performance degradation, but I didn't look into it in detail yet. The mixed precision training runs are still show a high idle time, but slow down for mirrored strategy isn't as significant as before. [Here are the TensorBoard profiles](https://github.com/tensorflow/tensorflow/files/5555716/logs.zip) for the measurements shown below:

version | FP32 | FP32 Mirrored | Keras FP16 | Keras FP16 Mirrored
--- | --- | --- | --- | --- 
2.4.0rc1 (`drop_remainder=True`) | 54s | **54s** | 34s | **35s**

**Comment by zongweiz** [comment]
@lgeiger Thanks for your information. Yes, we have tracked down the performance issue to a tf.cond inside Keras batch norm layer (to handle empty batches). Setting drop_remainder=True avoid empty/partial batches and workaround the problem. 
@rohan100jain is working on a fix and will give an update very soon

I think the reason we see more regression on FP16 is because: the above problem is more significant when the workload has higher overhead in launching GPU kernels, fp16 makes some compute kernels more efficient and makes the workload more kernel launch bound.

**Comment by lgeiger** [comment]
@zongweiz Thanks for looking into it. Looking forward to a fix.

**Comment by lgeiger** [comment]
@rohan100jain Do you have any updates whether a fix will make it into the 2.4 release?

**Comment by lgeiger** [comment]
@rohan100jain @zongweiz @jaingaurav Sorry for pinging you again. Do you have any updates on whether the fix will make it into the 2.4 stable release?

**Comment by guptapriya** [comment]
Hey @lgeiger i believe the fix did not make it to 2.4 unfortunately. @rohan100jain @zongweiz @goldiegadde  please correct if that is not the case.

**Comment by lgeiger** [comment]
> Hey @lgeiger i believe the fix did not make it to 2.4 unfortunately.

Thanks for the response, that's really unfortunate since the regression has been already there since 2.2. But at least we can now upgrade to 2.4 when enabling `drop_remainder=True` as a workaround until a fix lands.

**Comment by lgeiger** [comment]
@guptapriya has there been any progress on this issue? It would be good if a fix would make it into TF 2.5 since in our current workloads running on 4 GPUs we are seeing slowdowns of 30-80% due to this bug compared to TF 2.2 (using XLA makes this regression even more dramatic).

**Comment by guptapriya** [comment]
@lgeiger It looks like the last fix that was tried in November did not fix the issue and I don't see any other updates since then. Checking with @rohan100jain , will update when I know more.

**Comment by rohan100jain** [comment]
Apologies but I tried fixing it last year and that change had to be rolled back / didn't do what we expected to. We'll continue to work on it and get a fix out by 2.5

**Comment by lgeiger** [comment]
@rohan100jain Thanks for the update. Let me know when a fix lands and I am can rerun the benchmarks to verify.

**Comment by rohan100jain** [comment]
I'm sorry we looked into this issue and there isn't really any easy way of fixing this without rolling back a change (https://github.com/tensorflow/tensorflow/commit/f0d0485b0de521ff273c8d91acbb2fbabe57baa7) that enhances dtype coverage of our GPU ops and improves the consistency of Tensorflow in general. This issue has exposed some problems we need to fix with our device placement that we're planning to work on and will have an RFC for it. I'll therefore recommend that you continue to use the drop_remainder=True workaround for now.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
